File: python/백준/백트래킹/1759 암호 만들기.py
import sys

input = sys.stdin.readline
 
L, C = map(int, input().split())

alhabet = list(input().split())
alhabet.sort()  # 사전 순을 위해 미리 정렬하는 게 안전함
used_alhabet = [False] * C

vowels = {'a', 'e', 'i', 'o', 'u'}
answer = []

def back_tracking(arr, now_length):
    if now_length == L:
        # 모음/자음 조건 체크
        vowel_cnt = sum(1 for ch in arr if ch in vowels)
        consonant_cnt = L - vowel_cnt
        if vowel_cnt >= 1 and consonant_cnt >= 2:
            answer.append(''.join(arr))
        return

    for idx in range(C):
        if not used_alhabet[idx]:
            # 사전 순(오름차순) 유지
            if now_length == 0 or arr[now_length-1] < alhabet[idx]:
                used_alhabet[idx] = True
                arr.append(alhabet[idx])
                back_tracking(arr, now_length+1)
                arr.pop()
                used_alhabet[idx] = False

back_tracking([], 0)

# answer 자체가 이미 사전 순이지만, 혹시 몰라 정렬 한 번 더
answer.sort()

for comb in answer:
    print(comb)

File: cpp/백준/백트래킹/1759 암호 만들기.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <set>
using namespace std;

int L, C;
vector<string> alhabet;
vector<bool> used_alhabet;
set<char> vowels = {'a', 'e', 'i', 'o', 'u'};
vector<string> answer;

void back_tracking(vector<string>& arr, int now_length) {
    if (now_length == L) {
        int vowel_cnt = 0;
        for (auto& ch : arr) {
            if (vowels.count(ch[0])) vowel_cnt++;
        }
        int consonant_cnt = L - vowel_cnt;
        if (vowel_cnt >= 1 && consonant_cnt >= 2) {
            string s;
            for (auto& ch : arr) s += ch;
            answer.push_back(s);
        }
        return;
    }

    for (int idx = 0; idx < C; ++idx) {
        if (!used_alhabet[idx]) {
            if (now_length == 0 || arr[now_length - 1] < alhabet[idx]) {
                used_alhabet[idx] = true;
                arr.push_back(alhabet[idx]);
                back_tracking(arr, now_length + 1);
                arr.pop_back();
                used_alhabet[idx] = false;
            }
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> L >> C;
    alhabet.resize(C);
    for (int i = 0; i < C; ++i) cin >> alhabet[i];
    sort(alhabet.begin(), alhabet.end());
    used_alhabet.assign(C, false);

    vector<string> arr;
    back_tracking(arr, 0);

    sort(answer.begin(), answer.end());
    for (auto& comb : answer) cout << comb << "\n";
    return 0;
}
